### ISSUE TITLE
[BUG] cudf::byte_cast can return non-empty null values.

### ISSUE THREAD

**Issue by revans2** [rapidsai/cudf#11754]
**Describe the bug**
For short, int, and long values that are nulls the results will not point to an empty list.  Technically this does not violate the Arrow spec, but with changes going into CUDF for sorting of nested types, this can potentially cause incorrect results.

Not sure how critical this is, just because the list sorting is not done, and it may only really be a problem for lists of stricts, but I just don't know for sure.

**Steps/Code to reproduce bug**
Call cudf::byte_cast for an int column with `[null, 5]` in it. The result will look correct, but the first row will point to 4 bytes of data.

**Expected behavior**
The first column would point to the same offset for both the start and the end. (in this case 0).

**Comment by GregoryKimball** [comment]
@revans2 Does `byte_cast` create unsanitized lists? If I understand this function correctly, it can convert an `int64` column into a `List<int8>` column.

**Comment by bdice** [comment]
Correct, this function is creating unsanitized lists, which is a bug. [This logic](https://github.com/rapidsai/cudf/blob/9c06330363db4da99803a3728b8bf44f9829f0b9/cpp/src/reshape/byte_cast.cu#L66-L81) copies nonzero bytes even if the input element is null. The input's null mask is copied onto the output parent list column, but the null children are nonempty.
